HBASE-22547 Align the config keys and add document for offheap read in HBase Book.

openinx opened the pull request.

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 42 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
||| _ HBASE-21879 Compile Tests _ |
| +1 | mvninstall | 266 | HBASE-21879 passed |
| +1 | mvnsite | 1379 | HBASE-21879 passed |
| 0 | refguide | 481 | branch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 262 | the patch passed |
| +1 | mvnsite | 1109 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| 0 | refguide | 482 | patch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
||| _ Other Tests _ |
| +1 | asflicense | 12 | The patch does not generate ASF License warnings. |
| | | 4100 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/301 |
| Optional Tests |  dupname  asflicense  refguide  mvnsite  |
| uname | Linux 2b7af57aeb17 4.4.0-141-generic #167~14.04.1-Ubuntu SMP Mon Dec 10 13:20:24 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / a6e3d5ba47 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/1/artifact/out/branch-site/book.html |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/1/artifact/out/patch-site/book.html |
| Max. process+thread count | 86 (vs. ulimit of 10000) |
| modules | C: . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+For reducing the Java GC impact to P99/P999 RPC latency, HBase 2.x has made the offheap read and write path. The KV are
+allocated from the JVM offheap memory area, which won’t be garbage collected by JVM and need to be deallocated explicitly by
+upstream callers. On the write path, the request packet received from client will be allocated offheap and retained
+until those key values are successfully written to the WAL log and Memstore. The ConcurrentSkipListSet in Memstore does
```
key values -> Cells

openinx replied:
OK,  will replace all the kv/kvs with Cell/Cells. Thanks.

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+[[regionserver.offheap.overview]]
+== Overview
+
+For reducing the Java GC impact to P99/P999 RPC latency, HBase 2.x has made the offheap read and write path. The KV are
```
KV -> Cell

openinx replied:
Fine

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+For reducing the Java GC impact to P99/P999 RPC latency, HBase 2.x has made the offheap read and write path. The KV are
+allocated from the JVM offheap memory area, which won’t be garbage collected by JVM and need to be deallocated explicitly by
+upstream callers. On the write path, the request packet received from client will be allocated offheap and retained
+until those key values are successfully written to the WAL log and Memstore. The ConcurrentSkipListSet in Memstore does
```
ConcurrentSkipListSet -> Better we can say the in memory data structure in Memstore

openinx replied:
OK

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+not directly store the Cell data, but reference to cells, which are encoded in multiple Chunks in MSLAB,  this is easier
+to manage the offheap memory. Similarly, on the read path, we’ll try to read the BucketCache firstly, if the Cache
+misses, go to the HFile and read the corresponding block. The workflow: reading blocks from cache OR sending cells to
+client,  is basically not involved in heap memory allocations.
```
in heap -> on heap?

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+until those key values are successfully written to the WAL log and Memstore. The ConcurrentSkipListSet in Memstore does
+not directly store the Cell data, but reference to cells, which are encoded in multiple Chunks in MSLAB,  this is easier
+to manage the offheap memory. Similarly, on the read path, we’ll try to read the BucketCache firstly, if the Cache
+misses, go to the HFile and read the corresponding block. The workflow: reading blocks from cache OR sending cells to
```
cache OR sending ->  Some thing missing after OR?

openinx replied:
Actually, I mean: from reading block to sending cells to client, it's basically not involved on-heap memory allocation. Sorry about the writting.

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+upstream callers. On the write path, the request packet received from client will be allocated offheap and retained
+until those key values are successfully written to the WAL log and Memstore. The ConcurrentSkipListSet in Memstore does
+not directly store the Cell data, but reference to cells, which are encoded in multiple Chunks in MSLAB,  this is easier
+to manage the offheap memory. Similarly, on the read path, we’ll try to read the BucketCache firstly, if the Cache
```
We say cache firstly rather than say BC?

openinx replied:
OK

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+[[regionserver.offheap.readpath]]
+== Offheap read-path
+In HBase-2.0.0, link:https://issues.apache.org/jira/browse/HBASE-11425[HBASE-11425] changed the HBase read path so it
+could hold the read-data off-heap avoiding copying of cached data on to the java heap.
```
the read-data (from off heap BucketCache)

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+and pool them. Please make sure not to turn this OFF if you want end-to-end off-heaping in read path.
+If this pool is turned off, the server will create temp buffers on heap to accumulate the cell bytes and make a result cell block. This can impact the GC on a highly read loaded server.
+The user can tune this pool with respect to how many buffers are in the pool and what should be the size of each ByteBuffer.
+Use the config `hbase.ipc.server.reservoir.initial.buffer.size` to tune each of the buffer sizes. Default is 64 KB.
```
Now that its been changed to 65KB?

openinx replied:
For HBase2.0.0, the default value is still 64KB, while in HBase3.0.0, it will be 65KB. I think that's OK .

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+The user can tune this pool with respect to how many buffers are in the pool and what should be the size of each ByteBuffer.
+Use the config `hbase.ipc.server.reservoir.initial.buffer.size` to tune each of the buffer sizes. Default is 64 KB.
+
+When the read pattern is a random row read load and each of the rows are smaller in size compared to this 64 KB, try reducing this.
```
Not a comment on this patch.
Now we have the same BB pool to be used for reading the HFile blocks as well as for sending back the RPC responses.  For smaller row sizes we say here that users can tune the BB size to be lower. But that might impact the read from HFiles. I assume we don't change the block sizes in HFiles but keep the 64KB default.  So is this suggest us to have a large as well as small sized pool? Just an open Q for thinking.

openinx replied:
Yeah, If  user have a very small cell size, such as 100B, then shouldn't tune the buffer.size to 1KB or some thing smaller,  because of the block size.   Keep the 65KB  will be more better. So this  suggestion can be removed now I think.  Say even if keep a cell with 100B in a 65KB buffer, it won't have much impact, the buffer will put back to the pool and sending a cell  is quite short time.

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+In HBase-2.x, the RegionServer will still read block from HDFS to a temporary heap ByteBuffer and then flush to BucketCache's
+IOEngine asynchronously, finally it will be an offheap one.  We can still observe much GC pressure when cache hit ratio
+is not very high (such as cacheHitRatio ~ 60% ), so in link:https://issues.apache.org/jira/browse/HBASE-21879[HBASE-21879]
+we re-desgined the read path and made the HDFS block reading be offheap now. This feature will be available in HBASE-3.0.0.
```
re-desgined - typo

openinx replied:
OK

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+Firstly,  we introduced several configurations about the ByteBuffAllocator (which was abstracted to manage the memory application or release):
+
+1. hbase.ipc.server.reservoir.minimal.allocating.size: If the desired byte size is not less than this one, then it will be allocated as a pooled offheap ByteBuff, otherwise it will be allocated from heap directly because it
+is too wasting to allocate from pool with fixed-size ByteBuffers, default value is hbase.ipc.server.allocator.buffer.size/6.
```
hbase.ipc.server.allocator.buffer.size -> Newly added?  The existing one is 'hbase.ipc.server.reservoir.initial.max'.   If new u added, better avoid ipc?

openinx replied:
OK

openinx replied:
Oh, seems here can remove the ipc now for hbase.ipc.server.allocator.buffer.size

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+
+1. hbase.ipc.server.reservoir.minimal.allocating.size: If the desired byte size is not less than this one, then it will be allocated as a pooled offheap ByteBuff, otherwise it will be allocated from heap directly because it
+is too wasting to allocate from pool with fixed-size ByteBuffers, default value is hbase.ipc.server.allocator.buffer.size/6.
+2. hbase.ipc.server.allocator.max.buffer.count: The ByteBuffAllocator will have many fixed-size ByteBuffers inside which are composited as a pool, this config indicate how many buffers are there in the pool.
```
Any default for this?

openinx replied:
The default value is :    2MB * handlerCount / 65KB.

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+.Please make sure the buffer size is greater than your block size.
+
+We have the default block size=64KB, so almost all of the data block have a block size: 64KB + delta, whose delta is
+very small, depends on the size of last KeyValue. If we use the default hbase.ipc.server.allocator.buffer.size=64KB,
```
Pls change all ref to KV to Cell.

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+
+We have the default block size=64KB, so almost all of the data block have a block size: 64KB + delta, whose delta is
+very small, depends on the size of last KeyValue. If we use the default hbase.ipc.server.allocator.buffer.size=64KB,
+then each block will be allocated as a MultiByteBuff:  one 64KB DirectByteBuffer and one HeapByteBuffer with delta bytes,
```
Avoid referring as MBB which is internal stuff.  Say 2 BBs will be required then to read and carry the reading HFile block

openinx replied:
OK

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,146 @@
+
+.If disabled block cache, need to consider the index/bloom block size.
+
+Our default hfile.index.block.max.size is 128KB now, which means the index/bloom block size will be a little greater
```
Here we dont say at all why it is applicable for cache disabled case alone.  In a read work load, the cache disabled wont come at all I believe.  So better we dont say this here at all? Its kind of confusing for readers.

openinx replied:
Fine

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 27 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -0 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ HBASE-21879 Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for branch |
| +1 | mvninstall | 238 | HBASE-21879 passed |
| +1 | compile | 160 | HBASE-21879 passed |
| +1 | checkstyle | 128 | HBASE-21879 passed |
| +1 | mvnsite | 1115 | HBASE-21879 passed |
| 0 | refguide | 431 | branch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 262 | branch has no errors when building our shaded downstream artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 38 | HBASE-21879 passed |
| +1 | javadoc | 178 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 12 | Maven dependency ordering for patch |
| +1 | mvninstall | 234 | the patch passed |
| +1 | compile | 160 | the patch passed |
| +1 | javac | 160 | the patch passed |
| +1 | checkstyle | 129 | the patch passed |
| +1 | mvnsite | 1024 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| 0 | refguide | 446 | patch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 254 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 709 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 44 | the patch passed |
| +1 | javadoc | 176 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 12786 | root in the patch passed. |
| +1 | asflicense | 76 | The patch does not generate ASF License warnings. |
| | | 18961 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/301 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  refguide  mvnsite  |
| uname | Linux ae23840a33ac 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / 7840613bd1 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/2/artifact/out/branch-site/book.html |
| findbugs | v3.1.11 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/2/artifact/out/patch-site/book.html |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/2/testReport/ |
| Max. process+thread count | 5704 (vs. ulimit of 10000) |
| modules | C: hbase-common . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 28 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -0 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ HBASE-21879 Compile Tests _ |
| 0 | mvndep | 12 | Maven dependency ordering for branch |
| +1 | mvninstall | 232 | HBASE-21879 passed |
| +1 | compile | 171 | HBASE-21879 passed |
| +1 | checkstyle | 142 | HBASE-21879 passed |
| +1 | mvnsite | 1116 | HBASE-21879 passed |
| 0 | refguide | 452 | branch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 281 | branch has no errors when building our shaded downstream artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 41 | HBASE-21879 passed |
| +1 | javadoc | 183 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 13 | Maven dependency ordering for patch |
| +1 | mvninstall | 238 | the patch passed |
| +1 | compile | 176 | the patch passed |
| +1 | javac | 176 | the patch passed |
| +1 | checkstyle | 130 | the patch passed |
| +1 | mvnsite | 1107 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| 0 | refguide | 458 | patch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 266 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 759 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 53 | the patch passed |
| +1 | javadoc | 189 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 12484 | root in the patch passed. |
| +1 | asflicense | 78 | The patch does not generate ASF License warnings. |
| | | 18957 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/301 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  refguide  mvnsite  |
| uname | Linux e2d21f77dc81 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / 0d1228175b |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/3/artifact/out/branch-site/book.html |
| findbugs | v3.1.11 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/3/artifact/out/patch-site/book.html |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/3/testReport/ |
| Max. process+thread count | 5487 (vs. ulimit of 10000) |
| modules | C: hbase-common . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

anoopsjohn commented on hbase-common/src/main/java/org/apache/hadoop/hbase/io/ByteBuffAllocator.java:
```diff
@@ -114,7 +114,7 @@
 
   public static final String MIN_ALLOCATE_SIZE_KEY =
-      "hbase.ipc.server.reservoir.minimal.allocating.size";
+      "hbase.server.reservoir.minimal.allocating.size";
```
This is not part of this patch any more right?

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,168 @@
+
+Next thing to tune is the ByteBuffer pool on the RPC server side.
+The buffers from this pool will be used to accumulate the cell bytes and create a result cell block to send back to the client side.
+`hbase.ipc.server.reservoir.enabled` can be used to turn this pool ON or OFF. By default this pool is ON and available. HBase will create off heap ByteBuffers
```
Oh we had this config also using ipc. .. We could have deprecated this too the same way.  Sorry missed in last comment.

openinx replied:
Em..seems we can refactor this config as hbase.server.allocator.enabled... the hbase.ipc.server.reservoir.enabled is some confusing ...

anoopsjohn commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,168 @@
+and pool them. Please make sure not to turn this OFF if you want end-to-end off-heaping in read path.
+If this pool is turned off, the server will create temp buffers on heap to accumulate the cell bytes and make a result cell block. This can impact the GC on a highly read loaded server.
+The user can tune this pool with respect to how many buffers are in the pool and what should be the size of each ByteBuffer.
+Use the config `hbase.ipc.server.reservoir.initial.buffer.size` to tune each of the buffer sizes. Default is 64 KB for HBase2.x, while it will be changed to 65KB by default for HBase3.x
```
In all these new document, can we refer to new config at 1st place.  Also say the old one and is deprecated in 3.0

openinx replied:
That's OK.

openinx commented on src/main/asciidoc/_chapters/offheap_read_write.adoc:
```diff
@@ -0,0 +1,178 @@
+
+Next thing to tune is the ByteBuffer pool on the RPC server side:
+
+NOTE: the config keys which starts with prefix hbase.ipc.server.reservoir are deprecated in HBase3.x. If you are still
```
> In all these new document, can we refer to new config at 1st place. Also say the old one and is deprecated in 3.0
@anoopsjohn,  I've addressed your comment here.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 31 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -0 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ HBASE-21879 Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for branch |
| +1 | mvninstall | 252 | HBASE-21879 passed |
| +1 | compile | 173 | HBASE-21879 passed |
| +1 | checkstyle | 129 | HBASE-21879 passed |
| +1 | mvnsite | 1127 | HBASE-21879 passed |
| 0 | refguide | 465 | branch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 270 | branch has no errors when building our shaded downstream artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 37 | HBASE-21879 passed |
| +1 | javadoc | 175 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 14 | Maven dependency ordering for patch |
| +1 | mvninstall | 234 | the patch passed |
| +1 | compile | 161 | the patch passed |
| +1 | javac | 161 | the patch passed |
| +1 | checkstyle | 135 | the patch passed |
| +1 | mvnsite | 1034 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| 0 | refguide | 448 | patch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 265 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 732 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 47 | the patch passed |
| +1 | javadoc | 181 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 12273 | root in the patch failed. |
| +1 | asflicense | 79 | The patch does not generate ASF License warnings. |
| | | 18610 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestAdminShell |
|   | hadoop.hbase.client.TestTableShell |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/301 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  refguide  mvnsite  |
| uname | Linux e055d9ffdc2e 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / afaf7a9e35 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/5/artifact/out/branch-site/book.html |
| findbugs | v3.1.11 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/5/artifact/out/patch-site/book.html |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/5/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/5/testReport/ |
| Max. process+thread count | 5572 (vs. ulimit of 10000) |
| modules | C: hbase-common . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/5/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ HBASE-21879 Compile Tests _ |
| 0 | mvndep | 13 | Maven dependency ordering for branch |
| +1 | mvninstall | 256 | HBASE-21879 passed |
| +1 | compile | 171 | HBASE-21879 passed |
| +1 | checkstyle | 144 | HBASE-21879 passed |
| +1 | mvnsite | 1152 | HBASE-21879 passed |
| 0 | refguide | 483 | branch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 273 | branch has no errors when building our shaded downstream artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 245 | HBASE-21879 passed |
| +1 | javadoc | 206 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 11 | Maven dependency ordering for patch |
| +1 | mvninstall | 249 | the patch passed |
| +1 | compile | 172 | the patch passed |
| +1 | javac | 172 | the patch passed |
| -1 | checkstyle | 151 | root: The patch generated 2 new + 77 unchanged - 0 fixed = 79 total (was 77) |
| +1 | mvnsite | 1160 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| 0 | refguide | 487 | patch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 278 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 778 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 263 | the patch passed |
| +1 | javadoc | 208 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 17356 | root in the patch failed. |
| +1 | asflicense | 101 | The patch does not generate ASF License warnings. |
| | | 24519 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestTableShell |
|   | hadoop.hbase.client.TestAdminShell |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/301 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  refguide  mvnsite  |
| uname | Linux ab68cbc68874 4.4.0-141-generic #167~14.04.1-Ubuntu SMP Mon Dec 10 13:20:24 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / afaf7a9e35 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/6/artifact/out/branch-site/book.html |
| findbugs | v3.1.11 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/6/artifact/out/diff-checkstyle-root.txt |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/6/artifact/out/patch-site/book.html |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/6/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/6/testReport/ |
| Max. process+thread count | 5040 (vs. ulimit of 10000) |
| modules | C: hbase-common hbase-server . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/6/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

openinx commented:
Ping @anoopsjohn ,  any other concerns ?   I think it's the last subtask of HBASE-21879 until now.  Thanks.

anoopsjohn commented on hbase-common/src/main/java/org/apache/hadoop/hbase/io/ByteBuffAllocator.java:
```diff
@@ -68,10 +68,18 @@
   // default heap allocator, it will just allocate ByteBuffers from heap but wrapped by an ByteBuff.
   public static final ByteBuffAllocator HEAP = ByteBuffAllocator.createOnHeap();
 
+  public static final String ALLOCATOR_ENABLED_KEY = "hbase.server.allocator.enabled";
```
We should call it some way which can clear say if this is enabled, it is pooled stuff what we use. Anyways allocator will always be there in RS. It might be HEAP one.

openinx replied:
Sorry for the confusing, my fault.  let's define the 4 new config keys as: 

```
1. hbase.server.allocator.enabled -> hbase.server.allocator.pool.enabled
2. hbase.server.reservoir.minimal.allocating.size -> hbase.server.allocator.minimal.allocate.size
3. hbase.server.allocator.max.buffer.count
4. hbase.server.allocator.buffer.size
```
Will change all plances.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 10 | https://github.com/apache/hbase/pull/301 does not apply to HBASE-21879. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/7/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

openinx commented:
Updated the patch:
1.  Rename the comment message as: Align the config keys and add document for offheap read in HBase Book; 
2.  Reconsider the 4 new config keys. Say, defined as:
```
1. hbase.server.allocator.pool.enabled
2. hbase.server.allocator.minimal.allocate.size
3. hbase.server.allocator.max.buffer.count
4. hbase.server.allocator.buffer.size
```
Also checked all the places.
3. Fixed the checkstyle issues. 

@anoopsjohn , PTAL.

openinx commented:
Update the patch: 
1.  refactor the name ALLOCATOR_ENABLED_KEY as ALLOCATOR_POOL_ENABLED_KEY ;
2.  also add a separate log message for HBase3.x, like the following: 
```
Pool already reached its max capacity : XXX and no free buffers now. Consider increasing the value for 'hbase.ipc.server.reservoir.initial.max' ?
```
@anoopsjohn FYI.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 39 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ HBASE-21879 Compile Tests _ |
| 0 | mvndep | 28 | Maven dependency ordering for branch |
| +1 | mvninstall | 275 | HBASE-21879 passed |
| +1 | compile | 178 | HBASE-21879 passed |
| +1 | checkstyle | 146 | HBASE-21879 passed |
| +1 | mvnsite | 1178 | HBASE-21879 passed |
| 0 | refguide | 477 | branch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 275 | branch has no errors when building our shaded downstream artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 250 | HBASE-21879 passed |
| +1 | javadoc | 207 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 12 | Maven dependency ordering for patch |
| +1 | mvninstall | 250 | the patch passed |
| +1 | compile | 171 | the patch passed |
| +1 | javac | 171 | the patch passed |
| +1 | checkstyle | 154 | the patch passed |
| +1 | mvnsite | 1166 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| 0 | refguide | 490 | patch has no errors when building the reference guide. See footer for rendered docs, which you should manually inspect. |
| +1 | shadedjars | 277 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 793 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: . |
| +1 | findbugs | 269 | the patch passed |
| +1 | javadoc | 212 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 13069 | root in the patch failed. |
| +1 | asflicense | 86 | The patch does not generate ASF License warnings. |
| | | 20355 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/9/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/301 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  refguide  mvnsite  |
| uname | Linux 8d7c075235f1 4.4.0-141-generic #167~14.04.1-Ubuntu SMP Mon Dec 10 13:20:24 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / afaf7a9e35 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/9/artifact/out/branch-site/book.html |
| findbugs | v3.1.11 |
| refguide | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/9/artifact/out/patch-site/book.html |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/9/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/9/testReport/ |
| Max. process+thread count | 4622 (vs. ulimit of 10000) |
| modules | C: hbase-common hbase-server . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-301/9/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.
